docs(proposal): GitOps Promotion/Rollback Tool (#17506)
argoproj/argo-cd: docs/proposals/promotion-strategy.md

---
title: GitOps Promotion/Rollback Tool
authors:
  - "@crenshaw-dev"

>   - "@zaller"

Author: ```suggestion
  - "@zachaller"
```

![Diagram showing the promote-by-pr-merge strategy. Changes originate as PRs against the main (DRY) branch. A tool then generates PRs for each environment with the hydrated manifests. The tool merges the PRs in the designated order. Changes are immediately synced out to the cluster on PR merge.](images/promote-by-pr-merge.png)

#### Reversion

This proposal introduces a system for automated, atomic rollbacks via PRs. When a problem is detected, a PR is automatically opened against the affected environment's hydrated manifest branch reverting to the last known-healthy commit. The PR may be automatically merged, may wait for certain checks to pass, or may wait for a user to manually merge it.

### Concepts

#### Change

Every change is represented as a commit on a "DRY" branch in a git repo. The contents of the branch depend completely on the manifest hydration tool. For example, Kustomize may include kustomization.yaml and a few related manifest files; Helm may include values.yaml and supporting chart files; and so on.

#### Environment

> 

Reviewer: We generally advise against using branchers for environments, and instead recommend using folders. I would hate to enshrine branches in this way. I think you're probably familiar with the arguments for why that is but [here's a refresher](https://codefresh.io/blog/stop-using-branches-deploying-different-gitops-environments/).

Author: I need to clarify this section a bit... definitely agree "directories for environments" with respect to _dry_ manifests (e.g. your kustomize bases and overlays). But a separate branch for each environment's _hydrated_ manifests. So the user's "write" interface is directories. The "read" interface is branches.

It also has the nice side-benefit of letting you use SCMs' "compare" interface to compare environments' hydrated manifests. And you get a full history of the hydrated manifests for just one environment by looking at the branch's commits.

Author: From the blog post:

> If you follow a branch-based approach for your environments:
> 
> 1. You need to have 13 long living Git branches at all times.
> 2. You need 13 pull requests for promoting a single change across all environments.
> 3. You have a two dimensional promotion matrix with 5 steps upwards and 2-3 steps outwards.
> 4. The possibilities for wrong merges, configuration drift and ad-hoc changes is now non-trivial across all environment combinations.

I think these all apply to a system where the _user_ has to interact directly with those branches. The proposal would automate all those interactions.

Author: Exactly. Cheap and often has an awesome UI "for free."

The controller responsible for promoting a change through environments could create a CommitStatus indicating that dry commit 82bf3ac is `successful` because the associated hydrated commits have been committed to all environments and have all been deployed and marked as healthy.

The CommitStatus controller will synchronize the cluster state of a CommitStatus with the appropriate SCM (e.g. GitHub or Gitlab) via their commit APIs.

SCM support must be implemented via a plugin interface so that the community can contribute support for their SCMs of choice.

##### Relationship to other controllers

This CRD will be created by other controllers, an example being an Argo CD specific Health/Sync status controller which will stay updated in realtime for Health/Sync status of an App (or group of apps). There is another type of Pre-Commit type check that can be generated such as kustomzie build type checks these types share the same CommitStatus CRD where it is used determines its type. If it's on a ProposedCommit, it's a pre commit check. If it's just on the PromotionStrategy activeCommitStatuses, it's an active health check used for promotion.

##### Relationship to PromotionStrategy

> A CommitStatus by has no direct relationship to any PromotionStrategy (described below). To have any effect on promotion, the CommitStatus must be referenced (via a key selector) by a PromotionStrategy or a ProposedCommit which, in turn, is referenced by a PromotionStrategy. CommitStatus is meant to simply be an abstraction of the SCMs' commit status concepts. On its own, this abstraction should have no awareness of change promotion, GitOps, etc.

Reviewer 2: ```suggestion
A CommitStatus by itself has no direct relationship to any PromotionStrategy (described below). To have any effect on promotion, the CommitStatus must be referenced (via a key selector) by a PromotionStrategy or a ProposedCommit which, in turn, is referenced by a PromotionStrategy. CommitStatus is meant to simply be an abstraction of the SCMs' commit status concepts. On its own, this abstraction should have no awareness of change promotion, GitOps, etc.
```

> A CommitStatus by has no direct relationship to any PromotionStrategy (described below). To have any effect on promotion, the CommitStatus must be referenced (via a key selector) by a PromotionStrategy or a ProposedCommit which, in turn, is referenced by a PromotionStrategy. CommitStatus is meant to simply be an abstraction of the SCMs' commit strategy concepts. On its own, this abstraction should have no awareness of change promotion, GitOps, etc.

Reviewer 3: ```suggestion
A CommitStatus by has no direct relationship to any PromotionStrategy (described below). To have any effect on promotion, the CommitStatus must be referenced (via a key selector) by a PromotionStrategy or a ProposedCommit which, in turn, is referenced by a PromotionStrategy. CommitStatus is meant to simply be an abstraction of the SCMs' commit status concepts. On its own, this abstraction should have no awareness of change promotion, GitOps, etc.
```

##### Example

```yaml
apiVersion: commit-status.argoproj.io/v1alpha1
kind: CommitStatus
metadata:
  labels:
    commit-status.argoproj.io/key: argocd-health
spec:
  repoRef:
    name:
  # These fields roughly correspond to the fields available in the GitHub Checks API and the 
  # Gitlab Commit "pipeline status" API. 

>   sha: xyz321  # Corresponds to the hydrated SHA

Reviewer 4: Shouldn't this be a label as well to help the selection ?

Author: Possibly, but I think it might change quite often. Custom informer index on the user side might be the better option.

I've added a section explaining the mutability of this and the `repoRef` field.

  # Name is meant to be very brief. Corresponds to the `name` field in GitHub and the `name`/`context` field in Gitlab.
  name: hydration-successful
  # A longer description of the check. For GitHub, corresponds to the output.summary field (`name` will be reused for the output.title). For GitLab, corresponds to the `description` field.
  description:
  # The field is named `phase` to avoid implying it directly corresponds to either GitHub's `status` or GitLab's `state` field.
  phase: success # queued, in_progress, success, failure, or cancelled
  url: # For details about the status
status:
  plugin:
    github:
      checkRunId: # Optional int
    gitlab:
    # others can store their status here
```

#### PullRequest

A PullRequest represents a PR in some SCM.

> For the purposes of environment promotion, this resource will be created by the ProposedCommit controller.

Reviewer 3: ```suggestion
For the purposes of environment promotion, this resource will be created by the ProposedCommit and RevertCommit controllers.
```

Should these CRDs be under an `argoproj` group at all, or maybe something new?

### How should we represent repo access?

Argo CD's model of using Secrets to represent repositories has problems:

1) No fine-grained RBAC access: you either have access to Secrets or not, there's no way to specify "just repo secrets"
2) Argo-arbitrary interface: the Secret schema is meaningful only for Argo and not for other GitOps operators/tools
3) No schema validation: since Secrets have a relatively open-ended spec, writing and operating on repo secrets requires special care

This seems like a good opportunity to define a new repo interface which has a clear and specific API and can be easily used by other tools.

### How should the system handle change contention?

To keep things simple, we think new changes should always supersede old changes.

Consider a PromotionStrategy covering dev, test, and prod. If the test environment has an open PR against the hydrated branch (pending change), and someone pushes another commit to the main (DRY) branch, the PR will be updated to include the new changes, and the PR will wait for the dev environment to finish deploying and verifying that new commit.

> It's possible that a high-churn low environment could block merges on higher environments. So far, we think that's an acceptable tradeoff. Our recommendation will be that teams move as much of the change validation as possible to the pre-commit stage so that changes can be promoted relatively quickly through environments, and high-churn low environments won't block promotions to higher environments.

Reviewer 5: In order for the proposed system to be universally accepted by the community, please consider that many organizations test changes in one environment for arbitrarily long before promoting to the next one. 

Let's consider the following scenario. As far as I understand the proposal, the change to the dry branch is reviewed, approved and merged first. The change to the hydrated `dev` branch is then merged automatically as well, but the PRs to the hydrated `test` and `prod` branches remain open. After some time the user merges the PR to the hydrated `test` branch. After another few days the software is ready to be approved and merged to `prod`. But how many days have passed since reviewing the dry branch? In this hypothetical scenario it took 7 days. So the hydrated change to `prod` is being reviewed 7 days after the equivalent change to the dry branch. I find this odd. I believe the dry and hydrated changes should be approved by the same person at the same time. Especially that the dry change to one environment may not be the same as the dry change to another environment (for example, consider a change to an URL that is different in each environment). 

This is a fundamental concern I have with this proposal – the reviews, approvals and merges to the dry branch and the hydrated branches occur at completely different times. The person who approved the dry branch may not even be at work at the time the hydrated branches are being reviewed.

I find it confusing for other reasons as well: if I'm browsing the history of the changes to the dry branch, the merge time is divorced from the deployment request. That breaks one of the promises of GitOps: that `git log` roughly corresponds to the history of deployment requests.

I appreciate this proposal, however, especially the idea to make the PR the primary interface for promotions. So let me step back and consider what kind of reviews/approvals an organization may require between promotions to different environments:
1. An organization may require that the review and approval happens at the time of deployment to a given environment. For example, there may be a review and approval before deployment to `test` and another review/approval before deployment to `staging`. In this workflow, the user can have an arbitrarily long delay between deployments to `test` and `staging`.
2. An organization may require that the review and approval happens at the time of the first deployment in a  sequence of environments. For example, deployment to `staging` may require a review and approval of changes to both `staging` and `prod`, because developers should only be allowed to deploy to `staging` when they are ready to go to `prod` as well. In this workflow the software gets promoted from `staging` to `prod` within a couple of hours from deployment to staging.

Both of the workflows above are valid. So how can reviews and approvals look like in each case? Let me propose a different solution to the one you are proposing, but one that adheres to the same fundamental idea: that git and pull requests are the user interface.

In the first scenario the entire dry configuration for `test` and `staging` would be kept in separate directories in the dry branch, so that deploying to each environment involves a separate PR to the dry branch (one PR to the `test` directory of the dry branch and one PR to the `staging` directory of the dry branch). There is no need to have a PR to the rendered branch – the automation can just create a simple feature branch to be merged to the rendered branch (this also solves the PR rate limiting problem). Each PR can then be reviewed together with the corresponding rendered change at the same time as the dry change and by the same person. After merging the `test` PR, the user can wait arbitrarily long to raise the `staging` PR. The link to raise the `staging` PR is automatically generated and added as a comment to the `test` PR, so that the user can easily execute the promotion.

In the second scenario, the dry state of `staging` and `prod` should be kept in a single directory, i.e. using shared files (with the exception of some environment-specific overrides). That way a single PR produces rendered changes to both environments, which can be reviewed again at the same time as the dry change and by the same person. However, merging the PR only triggers deployment to `staging`. You might ask: how do you then approve promotions from staging to prod? By a special comment to the PR (such as `/promote`). The prod sync would wait until the user adds such a comment (this can be implemented in a number of ways, but I'll leave that discussion for another thread).

The advantages of the approaches above is that they don't separate out the reviews of the dry and hydrated manifests into steps that are done separately and a long time apart.

What do you think?

Author: ## Responses to concerns

> I believe the dry and hydrated changes should be approved by the same person at the same time.

100% agree. I think this proposal only makes sense when coupled with an automated diff bot which comments _all_ hydrated manifest changes on the dry PR for simultaneous review. That piece of the puzzle is discussed in the [ArgoCon talk](https://www.youtube.com/watch?v=p5EPKY3vM-E) but isn't yet sufficiently represented in this proposal.

> the reviews, approvals and merges to the dry branch and the hydrated branches occur at completely different times. The person who approved the dry branch may not even be at work at the time the hydrated branches are being reviewed.

I think we could assume that, for many environments, there will be no direct human review applied to the hydrated branch PR. We'd assume that the review was already done as part of the dry PR review. Of course, a user may choose to require a human review on the hydrated PR as an extra safety check. But I don't think that would be the default mode.

> if I'm browsing the history of the changes to the dry branch, the merge time is divorced from the deployment request

I believe we would try to make it clear that the dry commit history is a history of the developers' intent and the hydrated branch commit histories are the histories of the environments' states. The hydrated branch commits will be tied back to the dry commit history via README files and metadata files which contain auto-generated references to the associated dry commits.

## Responses to alternative proposals

### Proposal 1: developer-controlled wait between promotions

I'll answer the first proposal in two parts:

#### Part 1: two dry PRs

> Let me propose a different solution... the entire dry configuration for test and staging would be kept in separate directories in the dry branch, so that deploying to each environment involves a separate PR to the dry branch

I think the proposed system already accommodates this flow. If the developer wants to explicitly push only to `stage`, wait for a time, and then only push to `prod`, they can use the proposed system. After merging the first PR, only one promotion (hydrated) PR will be opened: one for the affected `stage` branch. After some time, the developer can open a second PR affecting only the `prod` environment, merge it, and a hydrated PR will be automatically opened/merged for the prod environment. Of course, making dry changes that affect all environments and then relying on automation to promote that change will be the preferred flow. But there is no technical limitation preventing them from taking a more manual approach. (The question of whether hydrated changes are represented as PRs or as direct commits is discussed in a separate thread, so I'll leave that discussion there.)

#### Part 2: auto-generating PR links

> The link to raise the staging PR is automatically generated and added as a comment to the test PR, so that the user can easily execute the promotion.

The link would open a PR against the stage directory of the dry branch or the stage hydrated branch? If the former, how would the system know what the source branch of the dry change is? If the latter, I feel like that's effectively the same as requiring a manual review on the auto-generated stage promotion PR against the hydrated branch.

### Proposal 2: `/promote`-based promotion PRs

I'm not sure how requiring a `/promote` comment on the dry PR is effectively different from requiring a manual approval on the hydrated PR for the target environment (via branch protection rules).

> The advantages of the approaches above is that they don't separate out the reviews of the dry and hydrated manifests into steps that are done separately and a long time apart.

I think both of the proposals above assume there is no automated diff presented to the developer on the dry PR. If I'm correct, we might not need to dive into the finer points of the alternative proposals, because your concerns may be resolved by the existence of the diff bot. :-)

Reviewer 5: Thanks for the thorough response :-) My comments are inline:

> > Proposal 1: developer-controlled wait between promotions
> > Part 2: auto-generating PR links
> > The link to raise the staging PR is automatically generated and added as a comment to the test PR, so that the user can easily execute the promotion.

> The link would open a PR against the stage directory of the dry branch or the stage hydrated branch? If the former, how would the system know what the source branch of the dry change is? If the latter, I feel like that's effectively the same as requiring a manual review on the auto-generated stage promotion PR against the hydrated branch.

It's the former – the link would open a PR against the `stage` directory of the dry branch. The system operates on the PR to the `test` environment, so it knows the branch associated with the PR. 

There are actually some nuances here. There are two use cases:
1. The developer deploys a number of changes to `test`, as they iteratively develop the change. So a number of PRs have already been deployed to `test`. Then the developer wants to deploy all of those changes to `staging`, so they effectively want to `copy` the current state of `test` to `staging`.
2. The developer is testing the newer version of the software in `test`, while production is running an older version. However, in the meantime, the developer needs to change the configuration in all environments: for example, they need to change a URL in the config map. This URL needs to change independently from the container image version running in each environment. So the developer raises a PR to `test` (the PR contains only the change to the URL) and effectively wants to apply the same patch to `staging` without applying all the other changes from `test` to `staging`. 

We have been discussing use case 1, but use case 2 requires additional discussion. 

Back when people used different branches for each environment, they could use the `git cherry-pick` command to perform the promotion in use case 2. But nowadays we use directories per environment (for good reasons), so we need an equivalent operation.

This is why I'm thinking about producing two links per environment for the user: one that raises a PR with the `copy/replace` semantics and the other with the `cherry-pick` semantics. The automation would retrieve the list of available environments, push two feature branches per environment and generate links in the form of: 
https://github.com/my_org/gitops-dry/pull/new/$BRANCH

> > Proposal 2: /promote-based promotion PRs

> I'm not sure how requiring a /promote comment on the dry PR is effectively different from requiring a manual approval on the hydrated PR for the target environment (via branch protection rules).

You are correct that the workflow is practically the same, but there are some implementational differences:

1. I believe it's a little confusing that the action of promoting from `staging` to `prod` involves approving a PR, because, as we previously discussed, the review of that PR actually gets done earlier in the process.
2. At the organization I'm currently contracting to, the review of the dry and rendered manifests is done by a developer, whereas the decision to go from `staging` to `prod` is made by a QA engineer. These are different people and I would rather avoid the confusion that the QA engineer supposedly reviews the rendered manifests, as this is how it's going to look like in the GitHub PR record.
3. There is also the question of access control: the QA engineer should only be granted the right to push a button to approve the promotion from `staging` to `prod`. Granting them the rights to approve rendered branches is going to look confusing and this is particularly important for organizations in regulated environments, which have to explain to auditors why certain user groups have certain rights. On the other hand, the access control of who can use the `/promote` comment can be implemented in the action that processes the comment and can be limited to that action only.
4. As discussed previously, using `/promote` avoids the PR creation rate limit problem. I'm not concerned about exceeding the rate limit when creating the PRs to the dry branch, because they are typically raised by humans and the limit applies to a user. It's creating PRs in the rendered branch that has the potential to exceed the limit, because these PRs would get raised by a bot. 

Finally, I'd like to note that there are potentially other ideas than `/promote`. @Oded-B suggested to me sometime ago that it might be possible to use Markdown checklists:

- [ ] promote to prod

Checking the box is an event that can trigger an action. However, I haven't tried this method myself and I haven't tested if the event contains the username of the user who checked/unchecked the box and if the action of checking and unchecking the box is properly recorded for auditing purposes. This would require some more investigation.

Author: > so they effectively want to copy the current state of test to staging.

How does the promotion system know what to copy in the dry branch? Consider the example change "convert Deployment to Argo Rollouts." For dev, you might want a very simple Rollout with just a couple steps. For test, you might want something slightly more cautious. And then for prod, you might have a very cautious Rollout configuration with many steps. The promotion system has no concept of a "change" that it can map to the dry manifests, so it cannot function by copying yaml from one environment directory to the other. It's up to the developer to express their "change" as yaml in all environment directories (or, if they choose, one at a time for extra control), and then it's up to the promotion system to deploy those changes according to the defined promotion rules. I'm not sure how we could implement a generalizable promotion system based on copying dry YAML from one environment directory to the other. (Codefresh's promotion system uses what amount to YAML "masks," implemented as json path expressions; I think that system is fraught with edge cases.)

> I believe it's a little confusing that the action of promoting from staging to prod involves approving a PR

It doesn't _have_ to involve approving a PR. You could implement a custom commit status check requiring that all prod promotions must be approved in some internal QA system. The status of that check would appear on the PR in the list of checks, and you could could add a "details" link out to your QA approval system for discoverability.

> These are different people and I would rather avoid the confusion that the QA engineer supposedly reviews the rendered manifests, as this is how it's going to look like in the GitHub PR record.

Understandable. I think the commit status checks-based solution resolves that concern.

> using /promote avoids the PR creation rate limit problem

Personally I think direct-commit is a better way to avoid the rate limit issue. But one could still use a `/promote` mechanism on rendered manifests instead of dry manifests.

    * No waiting around for the system to communicate its state
* There must be no intentional drift between hydrated branches and the cluster
* The tool must be usable by non-Argo GitOps tools - e.g. Flux should be able to integrate with this system without using esoteric adapters
* We must focus on providing a declarative, eventually-consistent experience rather than a "pipeline-driven" experience

### Non-Goals

* This is not an image updater; automated tag bumps must be done by another system, and this system can promote that change

## Proposal

#### Promotion

This proposal introduces a system of push-by-pr-merge. Instead of representing the change promotion state in some external system, it is represented as PRs against environment branches containing "hydrated" (or "rendered") manifests. The "wait" stages of the promotion process (performance tests, manual approvals, etc.) are represented as commit statuses blocking PR merges. The state of the entire system is visible in the git/SCM interface. This system is generalizable to many organizations, relying on near-ubiquitous SCM concepts, allowing the promotion tool to be open-sourced.

> ![Diagram showing the promote-by-pr-merge strategy. Changes originate as PRs against the main (DRY) branch. A tool then generates PRs for each environment with the hydrated manifests. The tool merges the PRs in the designated order. Changes are immediately synced out to the cluster on PR merge.](images/promote-by-pr-merge.png)

Reviewer 5: When implementing PR generation automation at a large enterprise, we've found that the GitHub rate limits for creating PRs are quite low. These rate limits aren't disclosed by GitHub and the [documentation](https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#about-secondary-rate-limits) states: _These secondary rate limits are subject to change without notice. You may also encounter a secondary rate limit for undisclosed reasons._ We've also found that the rate limits are higher for GitHub apps than GitHub service accounts, but these limits can also be easily exceeded with the rendered manifests pattern and hundreds of users. For that reason, I don't recommend designing a system that relies on automatically raising PRs. 

One way you can work around this limit is to have your automation generate a GitHub link that a human can click to generate a PR. Since the rate limit applies to a user, you are highly unlikely to exceed the rate limit that way. 

Another approach is to never use PRs for your rendered code, but rather generate feature branches and push them and merge them with your automation. Any reviews happen on PRs raised in the interface repository and only a human can raise a PR. And even in the interface repository, if a change needs to be automatically made and doesn't require approval, no PR is made, only a feature branch. This is the system we have implemented.

Reviewer 4: That's something we've encountered as well with renovate. And it's true that the conclusion was the same: if you need an automatic merge, probably a direct commit from the feature branch (where the checks happen) would be enough.

Author: Early on we discussed a mode which relies on merging without opening PRs (i.e. direct commit from a feature branch). I think for now we for simplicity we should start with PRs, find the rate-limits of that systems and any available workarounds, and then look into building out a direct-merge alternative.

The problem with direct-merge is that you lose the nice, familiar, visible interface. As the Argo CD UI evolves, we might find that it's sufficient and that the PR UI is redundant and unnecessary.

#### Git Push Access

This tool inherently must have git push access. We need to evaluate what systems/recommendations/documentation we need to make the system secure.

At a minimum, no component should have access to git push secrets which does not _need_ that access. For example, the controller which opens GitHub PRs should live in its own namespace with access to its push secrets, but the Argo CD ApplicationSet controller (for example) should not have access to those secrets.

We should also encourage users to use their SCM's branch protection features to limit environment branch push access to only the promotion/rollback tool. Developers shouldn't directly manipulate those branches.

#### Secret-Injection CMPs

This tool is not compatible with CMPs that inject secrets, because those secrets would then be pushed to git.

This tool will not be responsible for checking that no secrets are being written to git - that's the responsibility of the hydrator tool is creating ProposedCommits (for our purposes, that's Argo CD).

### Risks and Mitigations

### Upgrade / Downgrade Strategy

## Drawbacks

## Alternatives

> * Kargo
>     * Relies on many new concepts (Freight, Warehouse, etc.) instead of existing git/SCM concepts - i.e. requires training for developers to use
>     * Currently focuses on promote-by-sync and promote-by-push (which have drawbacks, as described in [Motivation](#motivation))
>     * Could be adapted to implement promote-by-pr-merge, but this is just one mode of many
> * Telefonistka

Reviewer 6: Some shameless promotion :)

```suggestion
* [Kargo](https://github.com/akuity/kargo)
    * Relies on many new concepts (Freight, Warehouse, etc.) instead of existing git/SCM concepts - i.e. requires training for developers to use
    * Currently focuses on promote-by-sync and promote-by-push (which have drawbacks, as described in [Motivation](#motivation))
    * Could be adapted to implement promote-by-pr-merge, but this is just one mode of many
* [Telefonistka](https://github.com/wayfair-incubator/telefonistka)
```

Author: We are in the business of promotion here ;-)
